File: day05/solution.ts
import { readFileSync } from "fs"
import { Mapping, AlmanacMap, range } from "./types"

function parseSeeds(line: string): number[] {
    return line.split(/\s+/).slice(1).map(s => parseInt(s))
}

function parseMap(line: string): Mapping {
    const [dest, source, length] = line.split(' ').map(v => parseInt(v))
    return new Mapping(source,dest,length)
}


function parseMaps(parts: string[]){
    return parts.map(p => {
        const [label, ...mappings] = p.split('\n')
        return new AlmanacMap(label, mappings.map(parseMap))
    })
}

function part1(parts: string[]){
    const seeds = parseSeeds(parts[0])
    const almanacMaps = parseMaps(parts.slice(1))
    let lowest = Number.MAX_SAFE_INTEGER
    seeds.forEach(seed => {
        let location = seed
        almanacMaps.forEach(am => {
            location = am.map(location)
        })        
        lowest = location < lowest ? location : lowest
    })

    console.log(`Part 1: ${lowest}`)
}

function part2(parts: string[]){
    const seeds = parseSeeds(parts[0])
    const almanacMaps = parseMaps(parts.slice(1))
    
    let lowest = Number.MAX_SAFE_INTEGER
    for(let i = 0; i < seeds.length; i += 2){
        let ranges: range[] = [[seeds[i], seeds[i] + seeds[i+1] - 1]]
        for(let am of almanacMaps){
            ranges = ranges.map(r => am.mapRange(r)).flat()
        }
        ranges.forEach(r => {
            lowest = lowest > r[0] ? r[0] : lowest
        })
    }

    console.log(`Part 2: ${lowest}`)
}

const parts = readFileSync(process.argv[2]).toString().trim().split("\n\n")
part1(parts)
part2(parts)

File: day05/solution.py
import sys

class Mapping:
    def __init__(self, source: int, dest: int, length: int):
        self.source = source
        self.dest = dest
        self.length = length        
    
    def map(self, source: int) -> int | None:
        diff = source - self.source
        if diff >= 0 and diff < self.length:
            return self.dest + diff
        return None

class MappingList:
    def __init__(self, mappings: list[Mapping]):
        self.mappings = mappings
    
    def map(self, source: int) -> int:
        for m in self.mappings:
            v = m.map(source)
            if v is not None:
                return v        
        return source

def parse_seeds(line: str) -> list[int]:
    return [int(s) for s in line.split()[1:]]

def parse_mapping(line: str) -> Mapping:
    dest, source, length = line.split()
    return Mapping(int(source), int(dest), int(length))

def parse_mappings(mstr: str) -> MappingList:
    return MappingList([parse_mapping(m) for m in mstr.split('\n')[1:]]) # skip first line, it's text

def parse_input(parts: list[str]):
    seeds, *mappings = parts
    seeds = parse_seeds(seeds)
    return seeds, [parse_mappings(m) for m in mappings]

def part1(parts: list[str]):
    seeds, mappings = parse_input(parts)
    locations = []
    for seed in seeds:
        val = seed
        for m in mappings:
            val = m.map(val)
        locations.append(val)
    
    print('Part 1:', min(locations))
                
        

if __name__ == '__main__':
    with open(sys.argv[1]) as file:
        parts = file.read().split('\n\n')
    part1(parts)